### PR TITLE
WIP: Stripe subscriptions

### PR DESCRIPTION
**Pull request by dzervas**
This MR adds a new DJango app, which allows the user to subscribe to Print-Nanny through Stripe.

## Deployment changes
The following commands are required for djstripe to initialize (not that even with the `ENABLE_SUBSCRIPTIONS` feature flag to `False`, the djstripe DJango app is installed).
```python
python manage.py djstripe_init_customers
python manage.py djstripe_sync_plans_from_stripe  # Omit if feature is disabled
```

## User side
After a user registers, a trial subscription starts. Its status can be seen under the user's email with a badge:
image

A new "Subscriptions" menu item is inserted on the left sidebar:
image

After clicking it, the history of the subscriptions is shown:
image

A device is required to be able to subscribe. after at least a device is registered for the user, the subscribe button appears:
image

Pressing it, reveals a modal where the user can select the recurrence of the subscription for the selected devices:
image

Pressing one of the available prices, the user is redirected to Stripe to pay the subscription, where optionally a promo code can be entered:
image

After which the subscription is shown in the history with the respective "Premium" badge under the user's email:
image

During various steps, the customer is also informed:
- Trial ending
- Subscription ending
- Payment failed
- Payment succeeded

## Development setup

After the **TEST** Stripe keys are added in the environment, all of the containers should get started and the aforementioned deployment steps should be followed.

I added a stripe webhook proxy container in the local docker-compose and it shows some steps that have to be followed (`docker-compose -f local.yml logs -f stripe` to see them).

To sync everything from stripe to the django models (customers, etc.) run: `docker-compose -f local.yml run --rm django bash -c "python manage.py djstripe_sync_models"`. This will take quite a lot of time keep in mind.

## Developer usage

There are 2 new methods in the `User` DJango model: `is_trialing` which returns True when a user has ANY active trial and `is_subscribed` which returns True when a user has ANY subscription (yes, even a trial one).

In other news, each subscription object (`djstripe.models.Subscription.objects.filter(customer=user.id)`) has metadata a map with a single key `devices` and value comma separated string with the IDs of the `OctoPrintDevice` that the subscription is about.

## Stripe product setup

A single product should be created and its prices should be "standard pricing" and "recurring". NO free trial should be set! It's implemented inside the `subscriptions.model` receiver!

### CODE REVIEW

File: print_nanny_webapp/subscriptions/views.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    queryset = djstripe.models.Subscription.objects.all()
+    slug_field = "id"
+    slug_url_kwarg = "id"
+    context_object_name = "subscription"
+
+
+# TODO: Maybe this needs to be a normal class? Maybe throw it above?
+# Might not be possible due to "requires_action" but could use REST API thing
+# TODO: Use crispy form
+def subscriptions_payment_intent_view_create(request: HttpRequest):
```
Comment (dzervas) [inline print_nanny_webapp/subscriptions/views.py:78]: @leigh-johnson this is the actual functional code. The above are more or less to be removed.
How I can write this better? more elegant? I was thinking about REST framework that is used elsewhere in the project, is that the "right" way?
If yes, I'm a bit puzzled with the serializers. From the way I think of it, I don't need one, am I right?

(ignore absurd code such as `Plan.objects.first()` for now)

File: print_nanny_webapp/templates/subscriptions/list.html [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+                        {% endif %}
+                    </td>
+                </tr>
+            {% endfor %}
+        </tbody>
+    </table>
+</div>
+
+<script>
+    window.addEventListener('load', function () {
```
Comment (leigh-johnson) [inline print_nanny_webapp/templates/subscriptions/list.html:133]: Can you drop a breadcrumb back to the original source code here? Thank you! https://github.com/dj-stripe/dj-stripe/blob/stable/2.4/tests/apps/example/templates/payment_intent.html

File: print_nanny_webapp/subscriptions/views.py (hunk 1 above)
Comment (leigh-johnson) [reply print_nanny_webapp/subscriptions/views.py:43]: Would it be easier to forget the idea of opening a modal, and instead render separate form/detail views? 
https://github.com/dj-stripe/dj-stripe/blob/stable/2.4/tests/apps/example/views.py

Comment (leigh-johnson) [reply print_nanny_webapp/subscriptions/views.py:43]: I think the logic here is pretty contained to the webapp, so I don't think it makes sense to expose to the outside world via REST API. A template view makes sense to me. :+1: 

A good litmus test for "does this functionality belong in the REST API?" is:

* Do I need to `pip install print-nanny-client` and interact with REST endpoints? 
* Do I need to `npm install print-nanny-client` and interact with the REST endpoints?

Since subscriptions are contained within the webapp, there isn't a good reason to add REST interfaces on top of that. 

Does this distinction make sense?

Comment (dzervas) [reply print_nanny_webapp/subscriptions/views.py:43]: after many hours over that `views.py`, I understood that actually everything is irrelevant apart from the `template_file` inside `PurchaseSubscriptionView` and the `create_payment_intent` function. The rest of the code is for the old way of sources or tokens for stripe - which we don't want, as it does not support 3D secure, etc. and is incompatible with European law, from what I got.

While I agree with you, why would someone use the REST API for that, we need to transfer the `payment_intent_id` value inside the same step.

Maybe we could split the steps in 3-4 forms & views, but I'm not sure and I think that it's much more work rather than just keeping the current ugly code.

BTW the modal is now implemented and it's looking awesome

Comment (leigh-johnson) [reply print_nanny_webapp/subscriptions/views.py:43]: Gotcha, thank you for digging deep into this and understanding what's needed! 🙇‍♀️

No worries about the first pass feeling inelegant. I'm always cool with a "make it work, then make it work more-better" approach. How're you feeling about edge case coverage? Are there any requirements remaining? 

:shipit: One quick way to deal with edge cases is to trace the request handler with Honeycomb. I think you can do something like...

I use tracing as a substitute for unit tests sometimes (or as a first step to writing them - study how it breaks in the wild)
```
import beeline

# A trace for this fn should be started automatically by HoneyComb middleware: https://github.com/bitsy-ai/octoprint-nanny-webapp/blob/main/print_nanny_webapp/middleware/honeycomb.py#L7

def subscriptions_payment_intent_view_create(request: HttpRequest):
    span = beeline.start_span(context={"request", request, "name": "parse_payment_intent_request" })
    

    if request.method != "POST":
        ctx = {"STRIPE_PUBLIC_KEY": djstripe.settings.STRIPE_PUBLIC_KEY}
        return TemplateResponse(request, SubscriptionsListView.template_name, ctx)

    intent = None
    data = json.loads(request.body)
    # TODO: Put this in the form
    data["plan"] = djstripe.models.Plan.objects.first().id
    beeline.finish_span(span)

    span = beeline.start_span(context={"request", request, "name": "some_other_payment_intent_logic" })
    # ... some other code  
    beeline.finish_span(span)

```

Comment (leigh-johnson) [reply print_nanny_webapp/subscriptions/views.py:43]: No offense intended, by the way! I have a lot of confidence in your software writing skills. 

I basically assume all code (including / ESPECIALLY my own) will break in colorful and surprising ways 😂 ¯\_(ツ)_/¯

Comment (dzervas) [reply print_nanny_webapp/subscriptions/views.py:43]: ESPECIALLY about payments, lets not get reassured by my potentially "majestic" software skills and lets see those tests pass (thank you for your kind words though! :) )

File: print_nanny_webapp/subscriptions/views.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+        # Populate template with current subscriptions and stripe public key
+        ctx["STRIPE_PUBLIC_KEY"] = djstripe.settings.STRIPE_PUBLIC_KEY
+        ctx["SUBSCRIPTIONS"] = customer.subscriptions.all().order_by('-created')
+        ctx["PRODUCTS"] = djstripe.models.Product.objects.filter(active=True)
+        for p in ctx["PRODUCTS"]:
+            p.prices_list = p.prices.filter(active=True)
+
+        try:
+            ctx["DEVICES"] = OctoPrintDevice.objects.filter(user=self.request.user)
```
Comment (leigh-johnson) [inline print_nanny_webapp/subscriptions/views.py:44]: `OctoPrintDevice.objects.filter` will return an empty list `[]` here. If you'd like to raise an `OctoPrintDevice.DoesNotExist` exception, use `OctoPrintDevice.objects.get`

```
$ docker-compose -f local.yml run --rm django python manage.py shell
Type 'copyright', 'credits' or 'license' for more information
IPython 7.21.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from django.apps import apps

In [2]: OctoPrintDevice = apps.get_model("remote_control", "OctoPrintDevice")
In [4]: OctoPrintDevice.objects.filter(user=100)
Out[4]: <QuerySet []>

In [5]: OctoPrintDevice.objects.get(user=100)
---------------------------------------------------------------------------
DoesNotExist                              Traceback (most recent call last)
<ipython-input-5-f4970364ee52> in <module>
----> 1 OctoPrintDevice.objects.get(user=100)

/usr/local/lib/python3.8/site-packages/django/db/models/manager.py in manager_method(self, *args, **kwargs)
     83         def create_method(name, method):
     84             def manager_method(self, *args, **kwargs):
---> 85                 return getattr(self.get_queryset(), name)(*args, **kwargs)
     86             manager_method.__name__ = method.__name__
     87             manager_method.__doc__ = method.__doc__

/usr/local/lib/python3.8/site-packages/django/db/models/query.py in get(self, *args, **kwargs)
    427             return clone._result_cache[0]
    428         if not num:
--> 429             raise self.model.DoesNotExist(
    430                 "%s matching query does not exist." %
    431                 self.model._meta.object_name

DoesNotExist: OctoPrintDevice matching query does not exist.
```

File: print_nanny_webapp/subscriptions/views.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    # message.attach_alternative(html_body, "text/html")
+    message.send()
+
+# Payment action required is handled during the payment step inside the view
+# There are both invoice.paid and invoice.payment_succeeded but docs lean on the first
+# https://stripe.com/docs/billing/subscriptions/webhooks#tracking
+# @webhooks.handler("charge.succeeded")
+@webhooks.handler("invoice.paid")
+def subscriptions_invoice_paid(event):
+    # TODO: When a trial was paid, do nothing
```
Comment (leigh-johnson) [inline print_nanny_webapp/subscriptions/views.py:154]: Reading between the lines, does a trial have a separate paid/unpaid status? :thinking: That would make sense, since some companies want to run paid or discounted trials.

File: print_nanny_webapp/templates/subscriptions/list.html [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        </tbody>
+    </table>
+</div>
+{% endblock content %}
+
+{% block extra_javascript %}
+{{ block.super }}
+<script src="https://js.stripe.com/v3/"></script>
+
+<!-- Source based on https://medium.com/@berat703/django-stripe-integration-with-using-dj-stripe-1df59d17fa73 -->
```
Comment (leigh-johnson) [inline print_nanny_webapp/templates/subscriptions/list.html:151]: Thanks for breadcrumbing a link to the blog post!

**Review by leigh-johnson — CHANGES_REQUESTED** [review]
Just one minor revision in SubscriptionsListView - this is looking good to me, thank you for powering through this! :raised_hands:

File: print_nanny_webapp/subscriptions/views.py (hunk 4 above)
Comment (dzervas) [reply print_nanny_webapp/subscriptions/views.py:154]: apparently not. I just searched my emails and I didn't get an invoice paid mail during account creation. Most probably leftover TODO from the `charge.succeeded` webhook

**Review by leigh-johnson — APPROVED** [review]
:clap: :shipit:
